src/box/txn_limbo.c

```diff
@@ -519,6 +528,10 @@ txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn)
 		limbo->confirmed_lsn = lsn;
 		vclock_follow(&limbo->confirmed_vclock, limbo->owner_id, lsn);
 	}
+	if (limbo->confirmed_entry->lsn <= lsn) {
+		limbo->confirmed_entry = &limbo->head_entry;
+		limbo->confirmed_entry_is_valid = false;
+	}
```

Comment:
Here this `lsn` is always `== limbo->confirmed_lsn`. Your invariant was that `confirmed_lsn >= confirmed_entry->lsn`, which makes this condition always true. I added ` else {assert(false);}` to this `if` and the tests passed. Can we drop this `if` then?

Reply:
I was wrong in that comment about the invariant (fixed). There should be an inverse inequality. `confirmed_lsn <= confirmed_entry->lsn`.
> Here this lsn is always == limbo->confirmed_lsn.

It seems that this is not true. Let's imagine that several fibers executed their transactions, then collected confirmations. At this point, WAL slowed down greatly and each of the fibers got stuck in `txn_limbo_write_confirm`. Then the confirmation for the first transaction was written to disk and called `txn_limbo_read_confirm` here lsn will be strictly less than `confirmed_entry->lsn`. I'll try to reproduce this later.
But I agree that this check can be simplified to `if (limbo->confirmed_entry->lsn == lsn)`.

Reply:
One more comment about this invariant `confirmed_lsn <= confirmed_entry->lsn`: 
I just found out that it is only valid if the following patch is applied:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index 458b3027f..97f50886e 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -708,13 +718,17 @@ txn_limbo_confirm(struct txn_limbo *limbo)
        assert(e->lsn != -1);
        int32_t k = (int32_t)vclock_size(&limbo->vclock)
                - replication_synchro_quorum;
-       int64_t confirm_lsn = (k < 0) ? 0 :
+       int64_t confirm_lsn1 = (k < 0) ? 0 :
                vclock_nth_element(&limbo->vclock, k);
-       assert(confirm_lsn != -1);
+       assert(confirm_lsn1 != -1);
+       int64_t confirm_lsn2 = -1;
        for (; !rlist_entry_is_head(e, &limbo->queue, in_queue);
-              e = rlist_next_entry(e, in_queue))
-               if (e->lsn == -1 || e->lsn > confirm_lsn)
+              e = rlist_next_entry(e, in_queue)) {
+               if (e->lsn == -1 || e->lsn > confirm_lsn1)
                        break;
+               if (txn_has_flag(e->txn, TXN_WAIT_ACK))
+                       confirm_lsn2 = e->lsn;
+       }
        limbo->confirmed_entry = rlist_prev_entry(e, in_queue);
        /*
         * We have not made sure that the transaction following confirmed_entry
@@ -722,15 +736,15 @@ txn_limbo_confirm(struct txn_limbo *limbo)
         * valid.
         */
        limbo->confirmed_entry_is_valid = false;
-       txn_limbo_write_confirm(limbo, confirm_lsn);
-       txn_limbo_read_confirm(limbo, confirm_lsn);
+       txn_limbo_write_confirm(limbo, confirm_lsn2);
+       txn_limbo_read_confirm(limbo, confirm_lsn2);
 }
 ```
I suggest not guaranteeing this invariant, but simply leaving this check as is.

Reply:
I read my comments a month later and it took me a while to understand what I meant :sweat_smile:. Therefore, I decided to write in a little more detail why, for example, the invariant would break without this patch. For myself and perhaps for those who will read this patch. The thing is that a situation is possible when the master has not yet assigned the lsn to the transaction (lsn = -1), while the transaction has already been replicated to a quorum and an ACK is obtained. It sounds counterintuitive and doesn't immediately come to mind.
_See comment below_:
```c
	for (; !rlist_entry_is_head(e, &limbo->queue, in_queue);
	       e = rlist_next_entry(e, in_queue))
		if (e->lsn == -1 || e->lsn > confirm_lsn)
			break;
	limbo->confirmed_entry = rlist_prev_entry(e, in_queue);
	/** 
	 * rlist_prev_entry(e, in_queue)->lsn (limbo->confirmed_entry)
	 * may be < confirm_lsn here in case master has not yet assigned
	 * lsn to the transaction (lsn = -1)
	 */
	limbo->confirmed_entry_is_valid = false;
	txn_limbo_write_confirm(limbo, confirm_lsn);
	txn_limbo_read_confirm(limbo, confirm_lsn);
```